nabs/simulators.py

```diff
@@ -0,0 +1,244 @@
+"""Simulation and validation functions for plans"""
+import itertools
+import logging
+import sys  # NOQA
+from contextlib import contextmanager
+from typing import Any, Generator, Iterator, List, Tuple
+
+from bluesky.simulators import check_limits
+
+logger = logging.getLogger(__name__)
+
+
+class ValidError(Exception):
+    pass
+
+
+def raiser(*args: Any, **kwargs: Any):
+    raise ValidError('forbidden method called')
+
+
+# Be wary of how you specify these, they are keyed based on
+# how they were imported.
+default_patches = [
+            "sys.modules['ophyd'].sim.SynAxis.set",
+            "sys.modules['ophyd'].signal.EpicsSignal.put",
+]
+
+
+@contextmanager
+def patch_sys_modules(
+    modules: List[str] = default_patches
+) -> Generator[Any, None, None]:
+    """
+    takes a list of module names as strings and stores them,
+    replaces them with a raiser, and replaces them after
+
+    Need to use exec/eval here due to pass-by-reference issues
```

Comment:
Can you expand a bit on what kind of issues you had? I just want to verify that we need to use these.

I don't immediately see why doing something like
```
from ophyd.signal import EpicsSignal
EpicsSignal.put = raiser
```
would fail to have the same effect.
For that matter, is going through `sys.modules` necessary?